docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+* Good, because our CI would only need to test the core bindings we maintain, and others can be tested by their own maintainers (for example, a `gecko-js` binding generator should be tested in `mozilla-central` and not here)
+* Good, because a release in one binding wouldn't have an impact on any other ones
+* Bad, because it complicates version compatibility, and we will need more assertions to guarantee it
+* Bad, because testability for might increase in complexity. Bindings not in this repository will not have access to the extensive fixtures and examples we have. We might need to expose those if take this option.
+* Bad because it increases the number of crates we manage. As we would manage the core crates for the bindings we depend on (i.e kotlin, swift, python, and soon a C++/gecko one). This will make the release process painful, but we can add automation and clear crate ownership to help.
+
+Overall this option is preferred because the tradeoffs are clear and manageable and because there is a raising need to create bindings for gecko-js, which can't be tested end-to-end without a complex build system.
+
+### **[Option 3] Remove `uniffi_bindgen`, replace it with `uniffi_scaffolding` and have each binding generator implementation fully manage the generation of bindings**
+
+* Good, because maintainability will be clear, and members of the community can opt to maintain their own binding generators
+* Good, because we make it clear that each binding generator is responsible for its own quality
+* Bad, because new binding generators can't take advantage of the prior art of the existing binding generators
+* Bad, because it will be hard to guarantee the quality of binding generators since they won't depend on any traits from `uniffi_bindgen`
+* Bad, because it complicates version compatibility, and we will need more assertions to guarantee it
+* Bad, because testability for might increase in complexity. Bindings not in this repository will not have access to the extensive fixtures and examples we have. We might need to expose those if take this option.
+* Bad because it increases the number of crates we manage. As we would manage the core crates for the bindings we depend on (i.e kotlin, swift, python, and soon a C++/gecko one). This will make the release process painful, but we can add automation and clear crate ownership to help.
+
+
+
+## Decision Outcome
+
+Chosen option:
+
+*  **[Option 2] Expose a trait in `uniffi_bindgen` that can be implemented by other crates, to separate the implementation of the bindings**
+
+### Changes Proposed
+#### The exposed trait for binding generators to implement
+`unffi_bindgen` would define a public trait, `BindingGenerator` that includes one main function:
+```rs
+/// Takes in the ComponentInterface and writes the bindings into directory `out_dir`
+fn write_bindings<P: AsRef<Path>>(&self, ci: &CompenentInterface, out_dir: P) -> Result<()>;
```

Comment:
```suggestion
fn write_bindings<P: AsRef<Path>>(&self, ci: &ComponentInterface, out_dir: P) -> Result<()>;
```

How would we handle configuration (`uniffi.toml`)?
Right now we _first_ need to parse the UDL to then load the configuration and adjust it according to the target binding.
If we just pass the configuration file path (or leave that up to the binding implementation fully) we lose control that it's all neatly inside a single uniffi.toml

Reply:
I like the idea of parsing `uniffi.toml`, pulling out any config in a section named after the target language, and passing that in as another argument here. It would be simpler for implementors and encourage consistency for end-users.

Reply:
Seems like the right way, but then you run into the issue that parsing `uniffi.toml` means parsing both the general config section and language-specific config section.  I guess `BindingGenerator` would have an associated type for the language-specific config, and we would use that somehow in a generic config parser function.